### PR TITLE
Feature/add common

### PR DESCRIPTION
**Pull request by iggi42**
a few things were not clarified:
 - is ignoring the dialyzer warnings okay?
 - should check? work recursive or should another function do that?

### CODE REVIEW

**Comment by iggi42** [comment]
also tests to check on the protocols need to be written, now that I see coverage.

File: apps/common/lib/image.ex [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  only `:data` OR `:url` should be set.
+  """
+  alias Common.Image
+
+  defstruct [:data, :url]
+  @typedoc "the typespec of an image"
+  @type t :: %Image{}
+
+  defimpl Common.Validation, for: Image do
```
Comment (skwerlman) [inline apps/common/lib/image.ex]: replacing this with a regular function that pattern matches against the input would be way simpler imo:
```elixir
def valid?(%Image{}), do: true
def valid?(_), do: false
```

You could also add guards to validate types (though you'd need to have multiple function heads for optional types since they might not be there)

This'd also solve the recursion question you had, since you can just enforce the struct type in the pattern match

File: .dialyzer_ignore.exs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  {":0:unknown_function Function Common.Validation.Float.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Function.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Integer.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.List.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Map.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.PID.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Port.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Reference.__impl__/1 does not exist."},
+  {":0:unknown_function Function Common.Validation.Tuple.__impl__/1 does not exist."}
+]
```
Comment (skwerlman) [inline .dialyzer_ignore.exs]: not a huge fan of these ignores. imo we should only be ignoring an issue if its caused by a dep. if its our fault, we should fix it instead

File: apps/common/lib/common.ex [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule Common do
+  @moduledoc """
+  Documentation for Common.
+  """
+
+  @typedoc """
+  A map representation of `Database.Schema.ModTag`.
+  """
+  @type tag :: String.t()
```
Comment (skwerlman) [inline apps/common/lib/common.ex]: is there a reason this doesn't have its own module? it'd be nice to be able to match on `%Tag{}` rather than it just being a string. the other advantage is we could move the tag mapping table out of the database cfg and into a real module

File: apps/common/lib/image.ex [hunk 4]
```diff
@@ -0,0 +1,18 @@
+defmodule Common.Image do
```
Comment (skwerlman) [inline apps/common/lib/image.ex]: imo the structs like image, mod, modfile and tag should be in a submodule, like `Common.Struct.*` (and the files in a corresponding folder)

File: apps/common/lib/utils.ex [hunk 5]
```diff
@@ -0,0 +1,14 @@
+defmodule Common.Utils do
+  @moduledoc "Module for utilities used across belethor."
+
+  @doc """
+  macro to correctly log debug messages.
+
+  Calls `Logger.debug` in a way that makes dialyzer happy and does impact performance in production.
+  """
+  defmacro debug(msg) do
```
Comment (skwerlman) [inline apps/common/lib/utils.ex:9]: can we get analogues for the other log levels?

File: apps/common/lib/utils.ex [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    quote do
+      :ok = Logger.debug(fn -> unquote(msg) end)
```
Comment (skwerlman) [inline apps/common/lib/utils.ex:11]: i _think_ this should be `_ =` not `:ok =` since it can technically return an error tuple

File: apps/common/test/common_test.exs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+defmodule CommonTest do
+  use ExUnit.Case
+  doctest Common
```
Comment (skwerlman) [inline apps/common/test/common_test.exs:3]: it'd be nice to have test cases for those `valid?` functions

**Review by skwerlman — CHANGES_REQUESTED** [review]

File: apps/common/test/common_test.exs (hunk 7 above)
Comment (iggi42) [reply apps/common/test/common_test.exs:3]: yeah, I am on it.
may be even use property testing?

File: apps/common/lib/utils.ex (hunk 6 above)
Comment (iggi42) [reply apps/common/lib/utils.ex:11]: depends should the code abort if the logging does not work?
with debug, I would say yes.

If you want to log an error and that logging fails, the code should abort / crash too.

File: apps/common/lib/utils.ex (hunk 5 above)
Comment (iggi42) [reply apps/common/lib/utils.ex:9]: yes

File: apps/common/lib/image.ex (hunk 4 above)
Comment (iggi42) [reply apps/common/lib/image.ex]: we can do that. I don't really care, we won't get that much structs.

File: apps/common/lib/common.ex (hunk 3 above)
Comment (iggi42) [reply apps/common/lib/common.ex]: sounds logical.
I thought one module for a type, that is nothing more than a string would be overkill.

File: .dialyzer_ignore.exs (hunk 2 above)
Comment (iggi42) [reply .dialyzer_ignore.exs]: yeah, they are quite ugly.
But I don't see the issue that the missing implementations could cause and I favor the protocols for validation.

File: apps/common/lib/image.ex (hunk 1 above)
Comment (iggi42) [reply apps/common/lib/image.ex]: The advantage of a protocol is that every module has it's own validation logic in it's .ex file.

File: apps/common/lib/image.ex [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  @moduledoc """
+  Represents a image.
+
+  only `:data` OR `:url` should be set.
+  """
+  alias Common.Image
+
+  defstruct [:data, :url]
```
Comment (iggi42) [inline apps/common/lib/image.ex]: maybe an struct or a map is not the right datatype to handle images.
It may be necessary in the database, but elsewhere `@type image :: {:data, Base64.t()} | {:url, Url.t()}` looks better.

File: apps/common/lib/image.ex (hunk 1 above)
Comment (skwerlman) [reply apps/common/lib/image.ex]: a struct module can also have functions tho? so you could validate by calling `image |> Image.valid?()` or `mod |> Mod.valid?()`.
instead of a protocol to enforce the existence of validation functions, use a behaviour.
the main advantage of a protocol is that it doesn't care about the data type, but when working with a database struct, you generally already know what data you're dealing with, so thats negated.

File: apps/common/test/common_test.exs (hunk 7 above)
Comment (skwerlman) [reply apps/common/test/common_test.exs:3]: property-based testing is usually better than unit testing where it works; imo unit testing should be mainly used for known corner cases

File: apps/common/lib/utils.ex (hunk 6 above)
Comment (skwerlman) [reply apps/common/lib/utils.ex:11]: in that case, use a with statement, and raise when it fails to match :ok. something like
```elixir
with :ok = Logger.debug(fn -> unquote(msg) end) do
  :ok
else
  _ -> raise "Failed to log message!"
end
```

File: apps/common/lib/image.ex (hunk 8 above)
Comment (skwerlman) [reply apps/common/lib/image.ex]: the reason i like a struct here is consistency with the other database types

File: apps/common/lib/image.ex (hunk 4 above)
Comment (skwerlman) [reply apps/common/lib/image.ex]: i just like to keep code relatively organized is all. like should be organized with like and all that

File: apps/common/lib/common.ex (hunk 3 above)
Comment (skwerlman) [reply apps/common/lib/common.ex]: this is just for the same reason i listed under image.ex; similar things should be implemented in similar ways

File: apps/common/lib/struct/mod_file.ex [hunk 9]
```diff
@@ -1,8 +1,8 @@
-defmodule Common.Modfile do
+defmodule Common.Mod.Modfile do
```
Comment (skwerlman) [inline apps/common/lib/struct/mod_file.ex]: shouldn't this be Struct not Mod?

Comment (iggi42) [reply apps/common/lib/struct/mod_file.ex]: aaah, yes. It should be a Mod

File: apps/common/lib/image.ex (hunk 8 above)
Comment (iggi42) [reply apps/common/lib/image.ex]: but they are not the same thing.
Choosing the datatype so that validation logic is not required is more important than consistency, imho.

Comment (skwerlman) [reply apps/common/lib/image.ex]: you'd still need validation logic for a tuple tho. either way you need to actively check the type of data; the atom only tells you which check to run, same as the key in a struct. but this way the validation is the same as it is for other database types
and for typing, you can specify the types of the data in the struct per-key:
```elixir
@type t() :: %Image{
  data: Base64.t(),
  url: Url.t()
}
```

Comment (iggi42) [reply apps/common/lib/image.ex]: well, you don't need to validate that only one of these attributes has a value and the other is nil.
If we use `%ImageUrl{} | %ImageData{}` or `{:data, Base64.t()} | {:url, Url.t()}` is "only" about a map is not justified for one value vs consistency with the other attributes.
I really dislike the idea that a data thing (be it a map, tuple or whatever) can have an invalid state, if we use maps/structs for everything is not that important.

Comment (skwerlman) [reply apps/common/lib/image.ex]: checking for a key is a p trivial check tho. and the cost of avoiding that check is a sane array of data types.
```elixir
keys = Map.keys(image)
valid_keys =
  if :data in keys do
     :url not in keys
  else
     :url in keys
  end
```

elixir is a dynamically typed language, so there's always going to be potential for invalid states for data. that's the whole point of the validation function; you can check whether data is good before doing anything with it.

**Review by skwerlman — APPROVED** [review]
lgtm :shipit:
